### PR TITLE
Report displays for macOS

### PR DESCRIPTION
**Pull request by dnfield**
https://github.com/flutter/flutter/issues/125939 and https://github.com/flutter/flutter/issues/125938 for macOS

fixes https://github.com/flutter/flutter/issues/121352

### CODE REVIEW

File: shell/platform/embedder/embedder.h
Comment (loic-sharma) [inline shell/platform/embedder/embedder.h:830]: Just to double-check, is `size_t` the best type for this? Platform views use `int64_t` for their identifiers.

**Comment by dnfield** [comment]
```dart
import 'dart:ui';

// import 'package:flutter/widgets.dart';
// import 'package:flutter/services.dart';

void main() {
  // WidgetsFlutterBinding.ensureInitialized();
  // SystemChrome.setPreferredOrientations([DeviceOrientation.portraitUp]);
  print(PlatformDispatcher.instance.displays.toList());
  print(PlatformDispatcher.instance.implicitView!.display);
  PlatformDispatcher.instance.onMetricsChanged = () {
    print(PlatformDispatcher.instance.displays.toList());
    print(PlatformDispatcher.instance.implicitView!.display);
  };
}
```

Is the simple Dart application I've been using to test this. In a default Flutter app it's working well to report display sizes when the window gets moved around to different monitors or the monitor configuration changes.

File: shell/platform/embedder/embedder.h
Comment (loic-sharma) [inline shell/platform/embedder/embedder.h:1683]: Just to double-check, the width/height types (`double`) are intentionally different from the window metric event's width/height types (`size_t`)?

File: shell/platform/embedder/embedder.cc
Comment (loic-sharma) [inline shell/platform/embedder/embedder.cc:3002]: To fix https://github.com/flutter/flutter/issues/121352 you'll need to do pointer math instead of array indexing. This is necessary as the embedder might not have all the members on the `FlutterEngineDisplay` struct.

For example, see how the pointer events are "indexed":

https://github.com/flutter/engine/blob/6dc9b28677b7d3e65594864347a2c1e3c1b19481/shell/platform/embedder/embedder.cc#L2250-L2251

File: shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm:794]: Obj-C style here and elsewhere: `YES` or `NO` not `TRUE` and `FALSE`.

**Review by cbracken — COMMENTED** [review]
Just a few comments.

File: shell/platform/embedder/embedder.h
Comment (dnfield) [reply shell/platform/embedder/embedder.h:830]: Switching for uint64_t

File: shell/platform/embedder/embedder.h
Comment (dnfield) [reply shell/platform/embedder/embedder.h:1683]: No. I'll change to size_t

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (dnfield) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:732]: Done. In the error case I'm just etting refresh_rate to zero.

File: shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm
Comment (dnfield) [reply shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm:794]: Done. This was copy pasted from another test.

File: shell/platform/embedder/embedder.cc
Comment (dnfield) [reply shell/platform/embedder/embedder.cc:3002]: Done

Comment (loic-sharma) [reply shell/platform/embedder/embedder.cc]: Looks good, thank you for fixing this!

File: shell/platform/embedder/embedder.h
Comment (loic-sharma) [reply shell/platform/embedder/embedder.h]: @dnfield I think this was supposed to be `int64_t` instead of `uint64_t`. Let me know if I missed something though!

**Review by loic-sharma — APPROVED** [review]
LGTM once you've double-checked `FlutterWindowMetricsEvent.display_id`'s type.

Please make sure to also get Chris's approval for macOS embedder changes and Chris's/Chinmay's approval for embedder API changes.

**Review by cbracken — APPROVED** [review]
lgtm

File: shell/platform/embedder/embedder.h
Comment (dnfield) [reply shell/platform/embedder/embedder.h]: Ah just saw this. I misread int64_t for uint. I'm not sure why this would be signed, but I can make it signed to avoid confusion with platform views in this API.

File: shell/platform/embedder/embedder.h
Comment (dnfield) [inline shell/platform/embedder/embedder.h:811]: @loic-sharma fyi

File: shell/platform/embedder/embedder.h
Comment (loic-sharma) [reply shell/platform/embedder/embedder.h]: Gotcha, that makes sense. Thanks for checking!

File: shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm
Comment (cbracken) [reply shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm]: Will do a quick blast through to catch the stragglers. Thanks for mentioning.

**Pull request merged**
